docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md

```diff
@@ -993,6 +994,74 @@ The `-fexceptions` option is also needed for C code that needs to interoperate w
+
+| Compiler Flag                                               | Supported since  | Description                                                                                                    |
+|:------------------------------------------------------------|:----------------:|:---------------------------------------------------------------------------------------------------------------|
+| <span id="-fvtable-verify">`-fvtable-verify=std`</span>     | GCC 4.9.4        | Enable run-time checks for C++ virtual function pointers corruption after shared libraries have been loaded  |
+| <span id="-fvtable-verify">`-fvtable-verify=preinit`</span> | GCC 4.9.4        | Enable run-time checks for C++ virtual function pointers corruption before shared libraries have been loaded |
+| <span id="-fvtable-verify">`-fvtable-verify=none`</span>    | GCC 4.9.4        | Disable enable run-time checks for C++ virtual function pointers corruption.                                 |
+
+#### Synopsis
+
+In object-oriented programming, object methods rely on dynamic dispatching to select the right virtual function to call at runtime. The vtable is the table of virtual function pointers linked to the code of the virtual function to be executed. The vtable can be overwritten by vtable hijacking attack. This option is used to check, at runtime, that each virtual function call is valid for its object type. This check prevents virtual function pointers from being corrupted or overwritten, and stops the application (SIGABRT) if the pointer integrity check fails. This virtual table pointer check is based on the data structure built at application start-up.
+
+This option has three choices[^gccvtv]:
+
+- `std` - Built after shared libraries have been loaded and initialized, vtable data imply to process check at normal object initialization time.
+- `preinit` - Built before shared libraries have been loaded and initialized, vtable data enable check before shared libraries are loaded to process check at early execution time.
+- `none` - turns off the vtable checks.
+
+[^gccvtv]: Team GCC, [Program Instrumentation Options: `-fvtable-verify`](https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html#index-fvtable-verify), GCC Manual, August 2013
+
+### Performance implications
+
+Caroline Tice played an important role in integrating the vtable verification (VTV) by submitting the commit of this feature in GCC 4.9[^gccvtvcommit]. She involved on articles that give information about performance penalty. She is describe performance penalty following 3 points:
+
+- Call verification: Performance penalty between 5% and 10%.
+- Object permission changes: Between 400% and 700% slowdown for permission changes per object file, while a performance loss of 320 ms is noticeable for permission changes per binary.
+- Virtual function hashtable size: Storage of virtual function hashtable imply a big waste of space.
```

Comment:
Another piece of feedback from the [C/C++ Compiler BP Guide Call in 2024-05-16](https://docs.google.com/document/d/1JY8FREBPCUUFpuv7-4B9EjeS2MLDpel0dbG5DFWrTns/edit#bookmark=id.cobg2blj7ukq) was that _"performance numbers should be attributed to specific benchmarks, 5% impact in Chrome may be acceptable, but 5% impact on SPEC benchmarks would generally be undesirable"_

Reply:
@thomasnyman @siddhesh The performance impact of the `-fvtable-verify` option is discussed in several articles.

In "Enforcing Forward-Edge Control-Flow Integrity in GCC & LLVM", a section is devoted to the performance impact of virtual table verification (7.1 VTV Performance - https://www.usenix.org/system/files/conference/usenixsecurity14/sec14-paper-tice.pdf#page=12). This article reports a performance impact on the "SPEC CPU2006 C++ benchmarks" of between 2.4% and 19.6% on omnetpp, astar and
xalancbmk. Four other benchmarks (povray, namd, soplex and dealII) showed no significant effect on performance. This article also focuses on reducing the performance penalty of the three affected benchmarks. They found that the performance penalty of xalancbmk fell from 19.2% to 10.8% using devirtualization and to 8.7% using static linking and devirtualization. They also noted that PGO optimization (flag -ripa) can reduce the number of virtual calls to be checked.

This information is presented in figures and graphs in the "Improving Function Pointer Security for Virtual Method Dispatches" slides on the GCC website (https://gcc.gnu.org/wiki/cauldron2012?action=AttachFile&do=get&target=cmtice.pdf) and in the slides relating to the article "Enforcing Forward-Edge Control-Flow Integrity in GCC & LLVM" (https://www.usenix.org/sites/default/files/conference/protected-files/sec14_slides_tice.pdf).

These numbers can add consistence to the section about performance impact. I also think that optimization that improving performance of code compiled with vtv also has a place in the section about performance impact (PGO, static linking and devirtualization) .

I will integrate these some of these numbers and optimizations into my pull request.

Regards,

Reply:
I summarized number from SPECS benchmarks in "Performance implications" subsection